### PR TITLE
Communication: Add TokenCredentialAddHostHeaderPolicy for TokenCredential Requests

### PR DESCRIPTION
**Pull request by melneubert**
Download request for CallingServerClient  created with TokenCredential depends on a host header for authentication. This was previously missing and requests were failing. To fix this, a new HttpPipelinePolicy was created to add this header to requests.

### CODE REVIEW

File: sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/CallingServerClientBuilder.java [hunk 1]
```diff
@@ -279,20 +283,25 @@ public CallingServerClientBuilder clientOptions(ClientOptions clientOptions) {
         return this;
     }
 
-    private HttpPipelinePolicy createHttpPipelineAuthPolicy() {
+    private List<HttpPipelinePolicy> createHttpPipelineAuthPolicies() {
```
Comment (chrwhit) [inline sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/CallingServerClientBuilder.java:286]: return types and method parameters should use interface types. This should be IList<>

Comment (cochi2) [reply sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/CallingServerClientBuilder.java:286]: This is Java, List is the interface :)

Comment (chrwhit) [reply sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/CallingServerClientBuilder.java:286]: Sorry, just realized this is Java. Check to see if they return List from other methods.

File: sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/CallingServerClientBuilder.java [hunk 2]
```diff
@@ -341,4 +350,8 @@ private HttpLogOptions getHttpLogOptions() {
 
         return httpLogOptions;
     }
+
+    private String getHostNameFromEndpoint() {
+        return endpoint.replace("https://", "");
```
Comment (cochi2) [inline sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/CallingServerClientBuilder.java:355]: are we 100% sure that the endpoint will always have `https`?

**Comment by cochi2** [comment]
Can we add a new download playback test using MI?

File: sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/CallingServerClientBuilder.java (hunk 2 above)
Comment (melneubert) [reply sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/CallingServerClientBuilder.java:355]: Good point, we probably also want to check for http. Anything else to look for?

**Comment by melneubert** [comment]
> Can we add a new download playback test using MI?

There is a download playback test using MI already. I've validated it passing by running it on my machine with a document id that was created recently.

File: sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/CallingServerClientBuilder.java (hunk 1 above)
Comment (melneubert) [reply sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/CallingServerClientBuilder.java:286]: Yep, I was using ArrayList as the concrete class. I see List in some method signatures for other areas of the calling server sdks (such as join call).

File: sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/CallingServerClientBuilder.java (hunk 2 above)
Comment (bharat-kalyan-namburi) [reply sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/CallingServerClientBuilder.java:355]: is there a lib in java like URL in C# that does this for us?

Comment (cochi2) [reply sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/CallingServerClientBuilder.java:355]: Yeah, I recommend you to do as Bharat suggests. Let the library do the heavy lifting.
There is URL library where you can create the URL from a String and call `getHost()` to get the host name.

Comment (melneubert) [reply sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/CallingServerClientBuilder.java]: Thanks, updated.

**Pull request merged**